### PR TITLE
Add Urboot bootloader support for MicroCore

### PR DESCRIPTION
**Pull request by MCUdude**
The latest version of MicroCore supports the new [Urboot bootloader](https://github.com/stefanrueger/urboot).
This PR allows setting fuses and burning the correct bootloader using PlatformIO.

The internal 9.6 and 4.8 MHz oscillator in the ATtiny13/A is notorious for being inaccurate. So much so that using a UART bootloader might become difficult. That's why MicroCore provides [various bootloader binaries](https://github.com/MCUdude/MicroCore/tree/master/avr/bootloaders/urboot) the bootloader script can choose from in order to use one that matches the user specification.

Specify the desired `f_cpu` and `f_cpu_error`, and the correct binary will be flashed.
There are a few rules though.
* There only exist pre-compiled bootloader files for the following internal oscillator options:
  - 9.6 MHz
  - 4.8 MHz
  - 1.2 MHz
  - 600 kHz
  - 128 kHz
* For each base frequency, the user can specify an error between -10 and 10 percent, in 1.25% steps.
* `f_cpu_error` is only useful when the internal oscillator option is selected

I'd like to hear what @valeros thinks about this. I'll update the official PlatformIO docs and the MicroCore docs to reflect the changes when/if this gets merged.

Note that I will also migrate over to Urboot for my other Arduino cores as well, but this may take some time and is not covered in this PR.

### CODE REVIEW

**Comment by MCUdude** [comment]
@valeros this is just a proposal, and I'd like to hear your thoughts regarding this. I'm open to feedback if you have ideas for new features or think things should be done differently. What's important is that the PlatformIO implementation is rock solid!

File: builder/bootloader.py [hunk 1]
```diff
@@ -54,6 +55,41 @@ def get_suitable_optiboot_binary(framework_dir, board_config):
     return bootloader_path.replace(".hex", "_BIGBOOT.hex")
 
 
+def get_suitable_urboot_binary(framework_dir, board_config):
+    mcu = board_config.get("build.mcu", "").lower()
+    f_cpu = re.sub("[^0-9]", "", board_config.get("build.f_cpu", "16000000L"))
```
Comment (valeros) [inline builder/bootloader.py:60]: Wouldn't it be easier to just drop the closing `L` by using list comprehension:
```python
board_config.get("build.f_cpu", "16000000L")[:-1]
```
Also, if we cast the `f_cpu` variable to integer in this line, we won't need a lot of `int(f_cpu)` calls below.

File: builder/bootloader.py [hunk 2]
```diff
@@ -64,9 +100,13 @@ def get_suitable_optiboot_binary(framework_dir, board_config):
 if core in ("MiniCore", "MegaCore", "MightyCore", "MajorCore"):
     if not isfile(bootloader_path):
         bootloader_path = get_suitable_optiboot_binary(framework_dir, board)
+elif core == "MicroCore":
+    if not isfile(bootloader_path):
+        bootloader_path = get_suitable_urboot_binary(framework_dir, board)
 else:
     if not isfile(bootloader_path):
         bootloader_path = join(framework_dir, "bootloaders", bootloader_path)
+        print(bootloader_path)
```
Comment (valeros) [inline builder/bootloader.py:109]: Is this leftover debug code? If no, maybe a bit more info added? Something like:
```python
"Using bootloader image: `%s`" % bootloader_path
```

File: builder/bootloader.py [hunk 3]
```diff
@@ -78,16 +118,20 @@ def get_suitable_optiboot_binary(framework_dir, board_config):
...
+    unlock_bits = board.get("bootloader.unlock_bits", "0x3F")
 
 env.Replace(
     BOOTUPLOADER="avrdude",
     BOOTUPLOADERFLAGS=[
         "-p",
         "$BOARD_MCU",
         "-C",
-        '"%s"'
+        "%s"
```
Comment (valeros) [inline builder/bootloader.py:134]: It would be great to separate concerns regarding added functionality. Do you mind moving this changes to a separate PR?

**Comment by valeros** [comment]
Hi @MCUdude ! Many thanks for the PR. Please review my comments below.

**Review by valeros — CHANGES_REQUESTED** [review]

File: builder/bootloader.py (hunk 1 above)
Comment (MCUdude) [reply builder/bootloader.py:60]: I'm not very good with python, so I don't know all the lever tricks. There may or may not be a capital `L` or `UL` at the of the number. If there is a more efficient way to remove trailing letters, I'm all ears!

File: builder/bootloader.py (hunk 3 above)
Comment (MCUdude) [reply builder/bootloader.py:134]: Sure! I'll remove the commit, rebase and submit a new PR when this PR is ready to be merged

File: builder/bootloader.py (hunk 1 above)
Comment (valeros) [reply builder/bootloader.py:60]: IIRC, there are no `UL` literals in PlatformIO dev-platforms, but anyway you can use `board_config.get("build.f_cpu", "16000000L").strip("UL")` instead of importing the regex module.

File: builder/bootloader.py (hunk 2 above)
Comment (MCUdude) [reply builder/bootloader.py:109]: It doesn't hurt, and it makes the whole process a little more transparent. BTW what's the deal with the \`` quotes in your example? Are these necessary? And how does they differ from the single quotes ' ' ?

Comment (valeros) [reply builder/bootloader.py:109]: > BTW what's the deal with the `` quotes in your example? Are these necessary? And how does they differ from the single quotes ' ' ?

Nothing special, just a habit came from writing markdown.

File: builder/bootloader.py (hunk 1 above)
Comment (MCUdude) [reply builder/bootloader.py]: I've just pushed a fix for this as well. Please review 🙂

**Comment by valeros** [comment]
Thanks for the PR, merged!
